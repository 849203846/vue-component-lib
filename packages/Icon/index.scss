@import "../style/global";

@include b(icon) {
  overflow: hidden;
  vertical-align: middle;
  fill: currentColor;

  @include size(1em);

  @each $name, $value in $--icon-animations {
    @include m($name) {
      animation: $value;
    }
  }
}
