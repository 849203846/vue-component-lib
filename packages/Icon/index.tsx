import sfc from '../utils/sfc';

export interface IIconComponent {
  Props: {
    name: string,
    color?: string,
    size?: string | number[] | string[],
    rotate?: boolean,
  }
  Events: {
    onClick?: (event: Event) => void;
  }
}

export type StyleType = {
  color?: string,
  width?: string,
  height?: string,
  'font-size'?: string,
}

const convert = (x: number | string) => (typeof x === 'number' ? `${x}px` : x);

const { def, bem } = sfc<IIconComponent>('icon');

export default def({
  props: {
    name: {
      type: String,
      required: true,
    },
    color: {
      type: String,
    },
    size: {
      type: [String, Array as () => number[] | string[]],
    },
    rotate: {
      type: Boolean,
      default: false,
    },
  },

  render() {
    const styles: StyleType = {
      color: this.color,
    };

    if (typeof this.size === 'string') {
      styles['font-size'] = this.size;
    }
    if (Array.isArray(this.size) && this.size.length > 0) {
      [styles.width, styles.height = convert(this.size[0])] = (this.size as []).map(convert);
    }

    return (
      <svg
        aria-hidden="true"
        class={bem({ rotate: this.rotate })}
        style={styles}
        onClick={() => { this.$emit('click'); }}
      >
        <use xlinkHref={this.name}></use>
      </svg>
    );
  },
});
