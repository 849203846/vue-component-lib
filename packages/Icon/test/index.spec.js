import { mount } from '@vue/test-utils';
import Icon from '..';

const iconName = 'icon';

describe('Component Icon', () => {
  it('props name', () => {
    const wrapper = mount(Icon, {
      propsData: { name: iconName },
    });
    expect(wrapper.classes('v-icon')).toBe(true);
    expect(wrapper.find('use').attributes('href')).toBe(iconName);
    expect(wrapper.element).toMatchSnapshot();
  });

  it('props rotate', () => {
    const wrapper = mount(Icon, {
      propsData: {
        name: iconName,
        rotate: true,
      },
    });
    expect(wrapper.classes()).toEqual(['v-icon', 'v-icon--rotate']);
  });

  it('props color', () => {
    const wrapper = mount(Icon, {
      propsData: {
        name: iconName,
        color: 'red',
      },
    });
    expect(wrapper.element.style.color).toBe('red');
  });

  it('props size: String', () => {
    const wrapper = mount(Icon, {
      propsData: {
        name: iconName,
        size: '16px',
      },
    });
    expect(wrapper.element.style.fontSize).toBe('16px');
  });

  it('props size: Array<number>', () => {
    const wrapper = mount(Icon, {
      propsData: {
        name: iconName,
        size: [10],
      },
    });
    expect(wrapper.element.style.width).toBe('10px');
    expect(wrapper.element.style.height).toBe('10px');
  });

  it('props size: Array<number> 2', () => {
    const wrapper = mount(Icon, {
      propsData: {
        name: iconName,
        size: [16, 20],
      },
    });

    expect(wrapper.element.style.width).toBe('16px');
    expect(wrapper.element.style.height).toBe('20px');
  });

  it('props size: Array<string>', () => {
    const wrapper = mount(Icon, {
      propsData: {
        name: iconName,
        size: ['10vw', '5vh'],
      },
    });

    expect(wrapper.element.style.width).toBe('10vw');
    expect(wrapper.element.style.height).toBe('5vh');
  });

  it('click event', () => {
    const wrapper = mount(Icon, {
      propsData: { name: iconName },
    });
    wrapper.trigger('click');
    expect(wrapper.emitted().click).toBeTruthy();
  });
});
