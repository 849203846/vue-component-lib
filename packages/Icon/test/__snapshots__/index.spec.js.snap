// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Component Icon props name 1`] = `
<svg
  aria-hidden="true"
  class="v-icon"
>
  <use
    xlink:href="icon"
  />
</svg>
`;
